File: backend/services/chat_service.py
from models import Chat, User
from database import db
from datetime import datetime, timedelta

# Simple anti-spam: track last message per user in-memory
_last_message = {}
COOLDOWN_SECONDS = 2

def can_send(user_id):
    key = user_id or 'anon'
    now = datetime.utcnow()
    last = _last_message.get(key)
    if last and (now - last).total_seconds() < COOLDOWN_SECONDS:
        return False
    _last_message[key] = now
    return True

def save_chat(user: User, username: str, avatar: str, message: str, room='lobby'):
    c = Chat(user_id=user.id if user else None, username=username, avatar=avatar, message=message, room=room)
    db.session.add(c)
    db.session.commit()
    return c

def get_history(room='lobby', limit=100):
    return Chat.query.filter_by(room=room).order_by(Chat.created_at.asc()).limit(limit).all()

File: backend/services/chatService.js
const { run, all } = require('../db');

const lastMessageAt = {}; // map user_id -> timestamp
const COOLDOWN = 1500;

function canSend(userId){
  const now = Date.now();
  const key = userId || 'anon';
  if(lastMessageAt[key] && now - lastMessageAt[key] < COOLDOWN) return false;
  lastMessageAt[key] = now;
  return true;
}

function saveChat({user_id, username, avatar, message, room='lobby'}){
  run('INSERT INTO chats (user_id, username, avatar, message, room) VALUES (?,?,?,?,?)', [user_id, username, avatar, message, room]);
  const id = this.lastID;
  return id;
}

function getHistory(room='lobby', limit=200){
  return all('SELECT * FROM chats WHERE room=? ORDER BY id DESC LIMIT ?', [room, limit]).reverse();
}

module.exports = { canSend, saveChat, getHistory };
